### PR TITLE
Implement token-based authN/authZ in apiserver

### PR DESCRIPTION
**Pull request by cbgbt**
**Issue number:** #80 #76

**Description of changes:**
For all write operations on nodes, the apiserver now checks that the requester
lives on the Node to which it is attempting to write. This means that
brupop agents will be incapable of changing the brn objects of other nodes,
or draining pods from nodes they are not running on.

This is done using the `TokenReview` API in kubernetes. An auth token provided by clients is used to authenticate that client's identity, and then ensure that the client lives on the node that it is attempting to augment.

**Testing done:**
Unit tests pass. I've also run the controller using this code, with successful node updates occurring.

Logs a succesfully authenticated request appear like so:

```
{"v":0,"name":"apiserver","msg":"[HTTP REQUEST - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-69fqw","pid":1,"time":"2021-11-06T03:17:40.461609704+00:00","target":"apiserver::telemetry","line":22,"file":"apiserver/src/telemetry.rs","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.user_agent":"","http.method":"POST","http.route":"/bottlerocket-node-resource","http.target":"/bottlerocket-node-resource","http.client_ip":"192.168.15.25:59276","otel.kind":"server","request_id":"bfa88277-ed00-42b0-9da0-2ded3bf98634","node_name":"ip-192-168-27-1.us-west-2.compute.internal"}
{"v":0,"name":"apiserver","msg":"[CHECK_REQUEST_AUTHORIZED - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-69fqw","pid":1,"time":"2021-11-06T03:17:40.461810565+00:00","target":"apiserver::auth","line":153,"file":"apiserver/src/auth.rs","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","k8s_audiences":"Some([\"brupop-apiserver\"])","namespace":"brupop-bottlerocket-aws","http.user_agent":"","http.method":"POST","http.route":"/bottlerocket-node-resource","http.target":"/bottlerocket-node-resource","http.client_ip":"192.168.15.25:59276","otel.kind":"server","request_id":"bfa88277-ed00-42b0-9da0-2ded3bf98634","node_name":"ip-192-168-27-1.us-west-2.compute.internal","node_selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-27-1.us-west-2.compute.internal\", node_uid: \"c17427b2-b3d7-4908-86de-9d389ae86391\" }"}
{"v":0,"name":"apiserver","msg":"[CHECK_REQUEST_AUTHORIZED - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-69fqw","pid":1,"time":"2021-11-06T03:17:40.464345025+00:00","target":"apiserver::auth","line":153,"file":"apiserver/src/auth.rs","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","http.method":"POST","otel.kind":"server","node_name":"ip-192-168-27-1.us-west-2.compute.internal","namespace":"brupop-bottlerocket-aws","elapsed_milliseconds":2,"http.user_agent":"","http.route":"/bottlerocket-node-resource","http.target":"/bottlerocket-node-resource","k8s_audiences":"Some([\"brupop-apiserver\"])","http.client_ip":"192.168.15.25:59276","request_id":"bfa88277-ed00-42b0-9da0-2ded3bf98634","node_selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-27-1.us-west-2.compute.internal\", node_uid: \"c17427b2-b3d7-4908-86de-9d389ae86391\" }"}
{"v":0,"name":"apiserver","msg":"[CREATE_NODE - START]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-69fqw","pid":1,"time":"2021-11-06T03:17:40.464378674+00:00","target":"models::node::client","line":146,"file":"models/src/node/client.rs","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-27-1.us-west-2.compute.internal\", node_uid: \"c17427b2-b3d7-4908-86de-9d389ae86391\" }","http.user_agent":"","http.method":"POST","http.route":"/bottlerocket-node-resource","http.target":"/bottlerocket-node-resource","http.client_ip":"192.168.15.25:59276","otel.kind":"server","request_id":"bfa88277-ed00-42b0-9da0-2ded3bf98634","node_name":"ip-192-168-27-1.us-west-2.compute.internal"}
{"v":0,"name":"apiserver","msg":"[CREATE_NODE - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-69fqw","pid":1,"time":"2021-11-06T03:17:40.470734253+00:00","target":"models::node::client","line":146,"file":"models/src/node/client.rs","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","elapsed_milliseconds":6,"selector":"BottlerocketNodeSelector { node_name: \"ip-192-168-27-1.us-west-2.compute.internal\", node_uid: \"c17427b2-b3d7-4908-86de-9d389ae86391\" }","http.user_agent":"","http.method":"POST","http.route":"/bottlerocket-node-resource","http.target":"/bottlerocket-node-resource","http.client_ip":"192.168.15.25:59276","otel.kind":"server","request_id":"bfa88277-ed00-42b0-9da0-2ded3bf98634","node_name":"ip-192-168-27-1.us-west-2.compute.internal"}
{"v":0,"name":"apiserver","msg":"[HTTP REQUEST - END]","level":30,"hostname":"brupop-apiserver-7c66b5c45f-69fqw","pid":1,"time":"2021-11-06T03:17:40.470779521+00:00","target":"apiserver::telemetry","line":22,"file":"apiserver/src/telemetry.rs","http.scheme":"http","http.host":"brupop-apiserver.brupop-bottlerocket-aws.svc.cluster.local","http.flavor":"1.1","otel.status_code":"OK","http.status_code":200,"elapsed_milliseconds":8,"http.user_agent":"","http.method":"POST","http.route":"/bottlerocket-node-resource","http.target":"/bottlerocket-node-resource","http.client_ip":"192.168.15.25:59276","otel.kind":"server","request_id":"bfa88277-ed00-42b0-9da0-2ded3bf98634","node_name":"ip-192-168-27-1.us-west-2.compute.internal"}
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: apiserver/src/error.rs [hunk 1]
```diff
@@ -26,6 +29,9 @@ pub enum Error {
     TracingConfiguration {
         source: tracing::subscriber::SetGlobalDefaultError,
     },
+
+    #[snafu(display("The kubernetes WATCH on Pod objects has failed."))]
```
Comment (bcressey) [inline apiserver/src/error.rs]: nit:
```suggestion
    #[snafu(display("The Kubernetes WATCH on Pod objects has failed."))]
```

File: apiserver/src/api.rs [hunk 2]
```diff
@@ -103,10 +116,43 @@ pub struct APIServerSettings<T: BottlerocketNodeClient> {
 /// Runs the apiserver using the given settings.
 pub async fn run_server<T: 'static + BottlerocketNodeClient>(
     settings: APIServerSettings<T>,
+    k8s_client: kube::Client,
 ) -> Result<()> {
     let server_port = settings.server_port;
-    HttpServer::new(move || {
+
+    // Set up a reflector to watch all kubernetes pods in the namespace.
+    // We use this information to authenticate requests write requests from brupop agents.
```
Comment (bcressey) [inline apiserver/src/api.rs]: nit:
```suggestion
    // We use this information to authenticate write requests from brupop agents.
```

File: apiserver/src/api.rs [hunk 3]
```diff
@@ -103,10 +116,43 @@ pub struct APIServerSettings<T: BottlerocketNodeClient> {
...
+        pod_store,
+        watcher(
+            pods,
+            ListParams::default().labels(&format!("{}={}", LABEL_COMPONENT, AGENT)),
+        ),
+    );
+    let drainer = try_flatten_touched(pod_reflector)
+        .filter_map(|x| async move { std::result::Result::ok(x) })
+        .for_each(|pod| {
+            event!(Level::TRACE, pod_name = %pod.name(), "Processed event for Pod");
```
Comment (bcressey) [inline apiserver/src/api.rs]: Should we log the kind of event we processed?

File: apiserver/src/api.rs [hunk 4]
```diff
@@ -264,3 +325,6 @@ mod tests {
         }
     }
 }
+
+#[cfg(test)]
+mod test {}
```
Comment (bcressey) [inline apiserver/src/api.rs]: What is this for?

File: apiserver/src/auth/authorizor.rs [hunk 5]
```diff
@@ -0,0 +1,438 @@
+//! This module provides abstractions for authenticating and authorizing requests from brupop agents to make changes to
+//! their Node's resources (including BottlerocketNode custom resources, or Draining their host Nodes of Pods.)
```
Comment (bcressey) [inline apiserver/src/auth/authorizor.rs]: nit: so it doesn't read as much like Victorian era writing:
```suggestion
//! the underlying Node's resources (including BottlerocketNode custom resources, or draining the Node of Pods.)
```

File: apiserver/src/auth/authorizor.rs [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+
+    use k8s_openapi::api::authentication::v1::UserInfo;
+    use k8s_openapi::api::core::v1::PodSpec;
+    use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
+    use kube::runtime::reflector;
+    use kube::runtime::watcher::Event;
+    use maplit::btreemap;
+
+    // use std::convert::TryFrom;
+    // use std::str::FromStr;
```
Comment (bcressey) [inline apiserver/src/auth/authorizor.rs]: nit:
```suggestion
```

File: apiserver/src/auth/authorizor.rs [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+        node_selector: &BottlerocketNodeSelector,
+        auth_token: &str,
+    ) -> Result<(), AuthorizationError>;
+}
+
+// The k8s TokenReview authenticator adds the pod name in the `extra` field of the UserInfo
+// provided with our TokenReview.status. This is the key to retrieve it.
+//
+// Unfortunately, the presence in the `extra` field means that this is non-standard, so it's possible
+// that brupop will not work with some implementations of a TokenReview server.
```
Comment (bcressey) [inline apiserver/src/auth/authorizor.rs]: "non-standard" sounds scary - it seems like this is guaranteed by the current Kubernetes API implementation, at any rate.

File: apiserver/src/auth/middleware.rs [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+                        &apiserver_headers.k8s_auth_token,
+                    )
+                    .await?;
+                fut.await
+            })
+        }
+    }
+}
+
+// TODO
```
Comment (bcressey) [inline apiserver/src/auth/middleware.rs]: What is left to do here?

**Review by bcressey — APPROVED** [review]

File: apiserver/src/auth/authorizor.rs (hunk 7 above)
Comment (cbgbt) [reply apiserver/src/auth/authorizor.rs]: Yes, same with audience-awareness. I don't anticipate having problems with upstream Kubernetes, but it does seem like there are hints that a nonstandard implementation could do something different.

File: apiserver/src/api.rs (hunk 3 above)
Comment (cbgbt) [reply apiserver/src/api.rs]: Turns out `try_flatten_touched()` only shows the flattened results of an event; however, I've changed the log message to output the spec and status of the resulting object.

**Comment by cbgbt** [comment]
^ Rebase atop `0.2.0`. The change before that addresses comments left by @bcressey.

**Review by bcressey — APPROVED** [review]

**Review by gthao313 — APPROVED** [review]

**Pull request merged**